File: src/schemas/metadata/MetadataTupleType.ts
import { ClassProperties } from "../../typings/ClassProperties";

import { IMetadataTupleType } from "./IMetadataTupleType";
import { Metadata } from "./Metadata";

export class MetadataTupleType {
  public readonly name: string;
  public readonly elements: Metadata[];

  public readonly index: number | null;
  public readonly recursive: boolean;
  public readonly nullables: boolean[];

  /**
   * @internal
   */
  public of_map?: boolean;

  /**
   * @internal
   */
  private constructor(props: ClassProperties<MetadataTupleType>) {
    this.name = props.name;
    this.elements = props.elements;
    this.index = props.index;
    this.recursive = props.recursive;
    this.nullables = props.nullables;
  }

  /**
   * @internal
   */
  public static _From_without_elements(
    props: Omit<IMetadataTupleType, "elements">,
  ): MetadataTupleType {
    return MetadataTupleType.create({
      name: props.name,
      index: props.index,
      elements: null!,
      recursive: props.recursive,
      nullables: props.nullables.slice(),
    });
  }

  public static create(
    props: ClassProperties<MetadataTupleType>,
  ): MetadataTupleType {
    return new MetadataTupleType(props);
  }

  public isRest(): boolean {
    return (
      this.elements.length > 0 &&
      this.elements[this.elements.length - 1]!.rest !== null
    );
  }

  public toJSON(): IMetadataTupleType {
    return {
      name: this.name,
      index: this.index,
      elements: this.elements.map((elem) => elem.toJSON()),
      recursive: this.recursive,
      nullables: this.nullables.slice(),
    };
  }
}

File: packages/typia/native/core/schemas/metadata/MetadataTupleType.go
package metadata

type IMetadataSchema_ITupleType struct {
	Name      string
	Elements  []*IMetadataSchema
	Index     *int
	Recursive bool
	Nullables []bool
}

type MetadataTupleType struct {
	Name      string
	Elements  []*MetadataSchema
	Index     *int
	Recursive bool
	Nullables []bool
	Of_map    *bool
}

func MetadataTupleType__From_without_elements(props IMetadataSchema_ITupleType) *MetadataTupleType {
	return MetadataTupleType_create(MetadataTupleType{
		Name:      props.Name,
		Index:     props.Index,
		Elements:  nil,
		Recursive: props.Recursive,
		Nullables: append([]bool{}, props.Nullables...),
	})
}

func MetadataTupleType_create(props MetadataTupleType) *MetadataTupleType {
	return &MetadataTupleType{
		Name:      props.Name,
		Elements:  props.Elements,
		Index:     props.Index,
		Recursive: props.Recursive,
		Nullables: append([]bool{}, props.Nullables...),
		Of_map:    props.Of_map,
	}
}

func (obj *MetadataTupleType) IsRest() bool {
	return len(obj.Elements) > 0 && obj.Elements[len(obj.Elements)-1].Rest != nil
}

func (obj *MetadataTupleType) ToJSON() IMetadataSchema_ITupleType {
	elements := make([]*IMetadataSchema, 0, len(obj.Elements))
	for _, elem := range obj.Elements {
		elements = append(elements, elem.ToJSON())
	}
	return IMetadataSchema_ITupleType{
		Name:      obj.Name,
		Index:     obj.Index,
		Elements:  elements,
		Recursive: obj.Recursive,
		Nullables: append([]bool{}, obj.Nullables...),
	}
}
